network/p2p/middleware.go

```diff
@@ -311,23 +311,25 @@ func (m *Middleware) start(ctx context.Context) error {
 		if err != nil {
 			return fmt.Errorf("failed to create peer manager: %w", err)
 		}
-
-		select {
-		case <-m.peerManager.Ready():
-			m.log.Debug().Msg("peer manager successfully started")
-		case <-time.After(30 * time.Second):
-			return fmt.Errorf("could not start peer manager")
-		}
 	}
 
 	return nil
 }
 
+// PeerManager returns the peer manager used by the middleware.
+// IMPORTANT: the peer manager is set when the middleware is started. Calling this method before
+// the middleware is started will return nil.
+func (m *Middleware) PeerManager() network.PeerManager {
+	return m.peerManager
+}
+
 // stop will end the execution of the middleware and wait for it to end.
 func (m *Middleware) stop() {
+	// TODO: does middleware need to stop the peer manager before the network?
+	// if not, we don't need to wait here since it will be stopped as a regular component
 	mgr, found := m.peerMgr()
 	if found {
-		// stops peer manager
+		// wait for peer manager to stop
```

Comment:
Will this be a problem? This may cause the middle ware and network stopping to block forever, since it waits for peer manager to stop, but peer manager stoppage is never initiated?

Reply:
yea, this could cause a problem if the middleware was ever stopped independently of the node. @yhassanzadeh13 do you know if the peer manager needs to be stopped before the libp2p node or does it not matter?

Reply:
@peterargue ideally the `PeerManager` should be stopped _before_ `LibP2PNode`. Otherwise, `PeerManager` may attempt on connecting to peers through libp2p's `Host` (which is encapsulated into `LibP2PNode`). If `LibP2PNode` has already been stopped, this connection attempt results in an error log. As this log is internal to libp2p, we don't have control over its level or content. So, except for the possibility of posting a misleading log, it seems there is no severe consequences. However, with the current state of architecture, it seems a safer approach to move `PeerManafger` as a member-field of the `LibP2PNode` instead of `Middleware` and let `LibP2PNode` handle the lifecycle of the `PeerManager` (i.e., `Start` and stop). It makes more sense as `PeerManager` does not seem as a dependency on `Middleware`. We can move running [`updateLoop`](https://github.com/onflow/flow-go/blob/master/network/p2p/peerManager.go#L107) into a separate `ComponentManager` worker, and then let `LibP2PNode` handle its lifecycle.

Reply:
Revising my comment: 

In the continuation of the concern posed at the `network/middleware.go` and also the comment above, it makes more sense to treat the `PeerManager` as a member field of the `LibP2PNode` (than `Middleware`) but with a different lifecycle management, i.e., we can expose `PeerManager`'s `Component` behavior as an exportable method on `LibP2PNode`: 

```
func (n *Node) PeerManagerComponent() component.Component {
	return n.peerManager
}
```

In this way, we can still enqueue the `PeerManager` as a component independently in `scafold` but without introducing any anti-pattern. 

Moreover, by moving `PeerManager` from `Middleware` to the `LibP2PNode` we yield a cleaner architecture regarding Separation of Concerns, as most of the dependencies of `PeerManager` belong to the `LibP2PNode` and in the current state of affair, `Middleware` is only in charge of the lifecycle management of the `PeerManager`. 

Also, it is fine to let `PeerManager` stop independently of the `Middleware`. As far as I checked, it only causes an internal error log on the libp2p's side when it attempts to update the peers while the `LibP2PNode` is down. As we do not expose the internal logger of the libp2p, it should not cause concern. So the `Middleware` stop logic will no longer need to check for the termination of the `PeerManager`.